core/templates/filters/translate.pipe.spec.ts

```diff
@@ -0,0 +1,143 @@
+      interpolateParams?: Object) {
+    const str = this.translations[key];
+    if (!str) {
+      return key;
+    }
+    return str;
+  }
+}
+
+describe('TranslatePipe', () => {
+  let pipe: TranslatePipe;
+  let translateService: TranslateService;
+  let changeDecRef: ChangeDetectorRef;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [
+        TranslatePipe
+      ],
+      providers: [
+        { provide: TranslateService, useClass: MockTranslateService },
+        { provide: ChangeDetectorRef, useClass: MockChangeDetectorRef}
+      ]
+    }).compileComponents();
+    translateService = TestBed.get(TranslateService);
+    changeDecRef = TestBed.get(ChangeDetectorRef);
+    pipe = new TranslatePipe(
+      translateService, changeDecRef, new UtilsService());
+  }));
+
+  it('should translate', () => {
+    expect(pipe.transform('I18n_t_1')).toBe('Hello');
+    expect(pipe.transform('I18n_t_2', {val: 'World'})).toBe('Hello <[val]>');
+    expect(pipe.transform('I18n_t_2', {val: 'World'})).toBe('Hello <[val]>');
+    expect(pipe.transform('I18n_t_3')).toBe('I18n_t_3');
+    expect(pipe.transform('')).toBe('');
+    translateService.onLangChange.emit({lang: 'en'});
+  });
+
+  // The sole purpose of this test is to cover ngOnDestroy.
+  it('should destroy subscriptions', () => {
+    pipe.ngOnDestroy();
+    expect(pipe).not.toBeDefined;
+    // Reintializing the pipe because jasmine tries to destroy the pipe.
+    // But since ngOnDestroy method has been called and the pipe is destroyed.
+    // And if the pipe is not reinitialized karma will raise an error saying
+    // "cannot call ngOnDestroy of undefined".
+    pipe = new TranslatePipe(
+      translateService, changeDecRef, new UtilsService());
+  });
+});
+
+// Adding the following two lines for reference.
+// I18n_rogue_1: '<script>alert(\'Oppia\');</script>Hello',
+// I18n_rogue_2: '<oppia-img>Me</oppia-img>Hola'
+@Component({
+  // eslint-disable-next-line no-multi-str, angular/no-inline-template
+  template: '<h1 [innerHTML] = "\'I18n_rogue_1\' | translate"></h1>\
+              <h2 [innerHTML] = "\'I18n_rogue_2\' | translate"></h2>'
+})
+class MockComponent { }
+
+describe('Angular', () => {
+  let component: MockComponent;
+  let fixture: ComponentFixture<MockComponent>;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [
+        TranslatePipe,
+        MockComponent
+      ],
+      providers: [
+        { provide: TranslateService, useClass: MockTranslateService },
+        { provide: ChangeDetectorRef, useClass: MockChangeDetectorRef}
+      ]
+    }).compileComponents();
+    fixture = TestBed.createComponent(MockComponent);
+    component = fixture.componentInstance;
+  }));
+
+  it('should sanitize translations', fakeAsync(() => {
+    const compiledComponent = fixture.debugElement.nativeElement;
+    // The DOM should be empty for now since the translations haven't
+    // been rendered yet
```

Comment:
I have already said this before, but end all comments with a period. Write in full sentences.